### PR TITLE
Abstracting Region and NAT ami mapping

### PR DESCRIPTION
**Pull request by amitaekbote**
Unit tests pass , more testing required before final review

### CODE REVIEW

File: gen/installer/aws.py [hunk 1]
```diff
@@ -54,6 +54,83 @@
         'id': 'ap-southeast-2'
     }]
 
+
+region_to_ami_map = [
+    {
+        'name': 'ap-northeast-1',
```
Comment (cmaloney) [inline gen/installer/aws.py]: `region`?

File: gen/installer/aws.py [hunk 2]
```diff
@@ -54,6 +54,83 @@
         'id': 'ap-southeast-2'
     }]
 
+
+region_to_ami_map = [
```
Comment (cmaloney) [inline gen/installer/aws.py]: as long as we're here want to fix https://dcosjira.atlassian.net/browse/DCOS-12?

File: gen/installer/aws.py [hunk 3]
```diff
@@ -178,6 +269,7 @@ def render_cloudformation_transform(cf_template, transform_func=lambda x: x, **k
         {k: transform_func(v) for k, v in kwds.items()}
     )
 
+    print(template_str)
```
Comment (cmaloney) [inline gen/installer/aws.py]: remove debug statement

File: gen/installer/aws.py (hunk 2 above)
Comment (amitaekbote) [inline gen/installer/aws.py]: should be straightforward, I will add it as a separate commit

**Comment by amitaekbote** [comment]
@cmaloney ready for review

File: gen/installer/aws.py [hunk 4]
```diff
@@ -138,6 +215,24 @@ def get_test_session(config=None):
     return release.call_matching_arguments(release.storage.aws.get_session, config, True)
 
 
+def gen_ami_mapping(mappings):
```
Comment (cmaloney) [inline gen/installer/aws.py]: nit: if we made region_to_ami_map an actual map ratherthan a list which looks like:
```
{
  'ap-northeast-1': {
    'coreos': '',
    'stable': '',
    'el7': '',
    'natami': ''
  }
}
```

then we could really easily test the ami mapping wihtout copying the AMI ids around:
```python
assert result["ap-northeast-1"] == {'stable': gen.installer.aws.region_to_ami_map['ap-northeast-1']['stable']}
```

which makes the test less fragile (every time someone updates the AMIs currently they have to touch this test which they'll never think of until CI pings them). It also is a lot simpler than what you have to do with the list thing for the same effect:
```python
ap_northeast_1_list = list(filter(lambda x: x['region'] == 'ap-northeast-1', gen.installer.aws.region_to_ami_map))
assert len(ap_northeast_1_list) == 1
assert result["ap-northeast-1"] == {'stable': ap_northeast_1_list[0]['stable']}
```

It also makes gen_ami_mapping a little simpler:
```python
def gen_ami_mapping(mappings):
    # create new dict with required mappings
    # all will have region by default
    for region, amis in region_to_ami_map.items():
        # Build the final dictionary
        final[region] = dict()
        for key in mappings:
          # TODO(cmaloney): natami needs to be mapped to the `default` key for 
          # now. Should update the templates to remove the need.
          final_key = 'default' if key == 'natami' else key
          final[region][final_key] = amis[key]

    # Returning -1:1 to cut the curly braces off the outside of the json dictionary so it formats
    # right into the templates.
    return json.dumps(final, indent=4, sort_keys=True)[1:-1]
```

Comment (cmaloney) [inline gen/installer/aws.py]: To build the final dictionary could go even more one line removing the manual iteration over mappings with a dictionary comprehsion (although not necessarily a good thing):
```python
final[region] = {'default' if key == 'natami' else key: amis[key] for key in mappings}
```

File: gen/test_installer_aws.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import json
+import gen.installer.aws
+
+
+def test_gen_aws_mapping():
+    result = json.loads("{" + gen.installer.aws.gen_ami_mapping({"stable"}) + "}")
+    # check number of regions
+    assert len(result) == 10
+    # check format of response
+    assert result["ap-northeast-1"] == {'stable': 'ami-965899f7'}
```
Comment (cmaloney) [inline gen/test_installer_aws.py]: the copying of the AMI here is going to make it so anyone updating the AMI mappings (Which we usually do once a release). Be really nice if we programatically grabbed it

**Comment by cmaloney** [comment]
:ship:

**Pull request merged**